### DISCUSSION TITLE
Possible memory leak in binary_server_asyncio.py ?

### DISCUSSION THREAD

**Discussion by chrisa84** [FreeOpcUa/opcua-asyncio#606] [General]
We use the opcua-asyncio library and have started encountering performance issues over recent months (unfortunately I am not sure specifically when this started). Our use case is that we have a python application exposing multiple OPC UA servers. We also have a Flask API which then uses the OPC UA client to connect to these servers in order to read/set nodes.

I have attached a flamegraph below (zipped) showing what we are seeing when this occurs. 

[cpu-flamegraph.zip](https://github.com/FreeOpcUa/opcua-asyncio/files/6629909/cpu-flamegraph.zip)

Of note is the number of references to this line: 
`_await_closing_tasks (asyncua/server/binary_server_asyncio.py:170)`

Looking at the commit history, there seem to have been a number of changes around this area after the latest 0.9.14 release which look like they could potentially fix the issue.

What are the current plans in terms of releasing these changes?

**Comment by oroulet** [comment]
I would like to release but there are large low level changes I am afraid of. If you can test master and tell me it works, then it will speed up release ;-)
Do you use custom structures?

**Reply by chrisa84** [reply]
Hi, thanks for the speedy reply. Hope I understand your question correctly - we are using exposing custom ExtensionObject which is essentially an array of a structure with 3 ints and a bool in python world.

**Reply by chrisa84** [reply]
I have built our app against the master version of the opcua-asyncio library and will test this out this afternoon (running our QA regression against it) and let you know if I observe any strange behaviour.

**Reply by oroulet** [reply]
thanks

**Reply by chrisa84** [reply]
Hi, unsure if this is of any use to you but I built our python against the master asyncua and we had a bunch of failures. I haven't spent any time looking into the detail - could be I need to update how we are using the library specifically for some of the changes - but I have attached the log just in case this useful.

Note that the "OPC_HVAC_ValvesTable_Struct" referenced in the logs is our ExtensionObject.

[asyncua master log.txt](https://github.com/FreeOpcUa/opcua-asyncio/files/6632003/asyncua.master.log.txt)

**Reply by oroulet** [reply]
OK. due to using datatypes, init methods are now automatically generated and a few argument that were written wrongly are now changed to how the spec defines them. In that case the argument of Variant called  dimensions is now written Dimensions.
I should mention it somewhere but unfornately I am not paid to maintain that code ;-)

**Reply by oroulet** [reply]
I should at least bump a lot the versoin number so people understand there has been changes...
